// postgis/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* postgis - MIT License - https://github.com/jsuyog2/postgis */

// src/queries/bbox.ts
function bbox(table, geom_column, srid, filter) {
  return `
    SELECT
      ST_Extent(ST_Transform(${geom_column}, ${srid})) as bbox
    FROM
      ${table}
    ${filter ? `WHERE ${filter}` : ""}
  `;
}

// src/queries/centroid.ts
function centroid(table, force_on_surface, geom_column, srid, filter) {
  const fn = force_on_surface ? "ST_PointOnSurface" : "ST_Centroid";
  return `
    SELECT
      ST_X(
        ST_Transform(
          ${fn}(
            ${geom_column}
          ), ${srid})
      ) as x,
      ST_Y(
        ST_Transform(
          ${fn}(
            ${geom_column}
          ), ${srid})
      ) as y
  
    FROM
      ${table}
  
    ${filter ? `WHERE ${filter}` : ""}
  `;
}

// src/queries/geobuf.ts
function geobuf(table, bounds, geom_column, columns, filter) {
  const bounds_value = bounds ? bounds.split(",").map(Number) : null;
  return `
    SELECT
      ST_AsGeobuf(q, 'geom')
  
    FROM
    (
  
      SELECT
        ST_Transform(${geom_column}, 4326) as geom
        ${columns ? `, ${columns}` : ""}
  
      FROM
        ${table}
        ${bounds_value ? `, (SELECT ST_SRID(${geom_column}) AS srid FROM ${table} WHERE ${geom_column} IS NOT NULL LIMIT 1) sq` : ""}
  
      ${filter || bounds_value ? "WHERE" : ""}
      ${filter ? `${filter}` : ""}
      ${filter && bounds_value ? "AND" : ""}
      ${bounds_value && bounds_value.length === 4 ? `${geom_column} &&
        ST_Transform(
          ST_MakeEnvelope(${bounds_value.join()}, 4326),
          srid
        )
        ` : ""}
      ${bounds_value && bounds_value.length === 3 ? `${geom_column} &&
        ST_Transform(
          ST_TileEnvelope(${bounds_value.join()}),
          srid
        )
        ` : ""}
  
    ) as q;
  
    `;
}

// src/queries/geojson.ts
function geojson(table, bounds, id_column, precision, geom_column, columns, filter) {
  const bounds_value = bounds ? bounds.split(",").map(Number) : null;
  return `
      SELECT
        jsonb_build_object(
          'type',       'Feature',
          ${id_column ? `'id', ${id_column},` : ""}
          'geometry',   ST_AsGeoJSON(geom, ${parseInt(String(precision), 10)})::jsonb,
          'properties', to_jsonb( subq.* ) - 'geom' ${id_column ? `- '${id_column}'` : ""}
        ) AS geojson
  
      FROM (
        SELECT
          ST_Transform(${geom_column}, 4326) as geom
          ${columns ? `, ${columns}` : ""}
          ${id_column ? `, ${id_column}` : ""}
        FROM
          ${table},
          (SELECT ST_SRID(${geom_column}) AS srid FROM ${table} WHERE ${geom_column} IS NOT NULL LIMIT 1) a
        ${filter || bounds_value ? "WHERE" : ""}
          ${filter ? `${filter}` : ""}
          ${filter && bounds_value ? "AND" : ""}
          ${bounds_value && bounds_value.length === 4 ? `${geom_column} &&
            ST_Transform(
              ST_MakeEnvelope(${bounds_value.join()}, 4326),
              srid
            )
            ` : ""}
          ${bounds_value && bounds_value.length === 3 ? `${geom_column} &&
            ST_Transform(
              ST_TileEnvelope(${bounds_value.join()}),
              srid
            )
            ` : ""}
      ) as subq
    `;
}

// src/queries/intersect_feature.ts
function intersect_feature(table_from, table_to, columns, distance, geom_column_from, geom_column_to, filter, sort, limit) {
  return `
   SELECT
    ${columns}

  FROM
    ${table_from},
    ${table_to}

  WHERE
    ST_DWithin(
      ${table_from}.${geom_column_from},
      ${table_to}.${geom_column_to},
      ${distance}
    )
    ${filter ? `AND ${filter}` : ""}

  ${sort ? `ORDER BY ${sort}` : ""}

  ${limit ? `LIMIT ${limit}` : ""}
  `;
}

// src/utils/validation.ts
function parsePoint(point) {
  if (typeof point !== "string") {
    throw new TypeError(`[postgis] Point must be a string, got: ${typeof point}`);
  }
  const match = point.match(/^(-?\d+\.?\d*),(-?\d+\.?\d*),(\d{4,5})$/);
  if (!match) {
    throw new Error(
      `[postgis] Invalid point format "${point}". Expected "x,y,srid" (e.g. "73.5,14.9,4326").`
    );
  }
  const [, x, y, srid] = match;
  return { x, y, srid };
}

// src/queries/intersect_point.ts
function intersect_point(table, point, columns, distance, geom_column, filter, sort, limit) {
  const { x, y, srid } = parsePoint(point);
  return `
  SELECT
    ${columns}

  FROM
    ${table}

  WHERE
    ST_DWithin(
      ${geom_column},
      ST_Transform(
        st_setsrid(
           st_makepoint(${x}, ${y}),
          ${srid}
        ),
        (SELECT ST_SRID(${geom_column}) FROM ${table} LIMIT 1)
      ),
      ${distance}
    )
    ${filter ? `AND ${filter}` : ""}

  ${sort ? `ORDER BY ${sort}` : ""}

  ${limit ? `LIMIT ${limit}` : ""}
  `;
}

// src/queries/list_columns.ts
function list_columns(table) {
  return `
    SELECT
    attname as field_name,
        typname as field_type

    FROM
    pg_namespace, pg_attribute, pg_type, pg_class

    WHERE
    pg_type.oid = atttypid AND
    pg_class.oid = attrelid AND
    relnamespace = pg_namespace.oid AND
    attnum >= 1 AND
    relname = '${table}'
        `;
}

// src/queries/list_tables.ts
function list_tables(filter) {
  return `
    SELECT
      i.table_name,
      i.table_type,
      g.f_geometry_column as geometry_column,
      g.coord_dimension,
      g.srid,
      g.type
    FROM
      information_schema.tables i
    LEFT JOIN geometry_columns g
    ON i.table_name = g.f_table_name
    INNER JOIN information_schema.table_privileges p
    ON i.table_name = p.table_name
    AND p.grantee in (current_user, 'PUBLIC')
    AND p.privilege_type = 'SELECT'
    WHERE
    i.table_schema not in  ('pg_catalog', 'information_schema')
  
      -- Optional where filter
      ${filter ? `and ${filter}` : ""}
  
    ORDER BY table_name
    `;
}

// src/queries/mvt.ts
function mvt(table, x, y, z, columns, id_column, geom_column = "geom", filter) {
  return `
      WITH mvtgeom as (
        SELECT
          ST_AsMVTGeom (
            ST_Transform(${geom_column}, 3857),
            ST_TileEnvelope(${z}, ${x}, ${y})
          ) as geom
          ${columns ? `, ${columns}` : ""}
          ${id_column ? `, ${id_column}` : ""}
        FROM
          ${table},
          (SELECT ST_SRID(${geom_column}) AS srid FROM ${table} WHERE ${geom_column} IS NOT NULL LIMIT 1) a
        WHERE
          ST_Intersects(
            ${geom_column},
            ST_Transform(
              ST_TileEnvelope(${z}, ${x}, ${y}),
              srid
            )
          )
  
          ${filter ? ` AND ${filter}` : ""}
      )
      SELECT ST_AsMVT(mvtgeom.*, '${table}', 4096, 'geom' ${id_column ? `, '${id_column}'` : ""}) AS mvt from mvtgeom;
    `;
}

// src/queries/nearest.ts
function nearest(table, point, columns, geom_column, filter, limit = 10) {
  const { x, y, srid } = parsePoint(point);
  return `
    SELECT
      ${columns},
      ST_Distance(
        ST_Transform(
          st_setsrid( st_makepoint(${x}, ${y}), ${srid} ),
          (SELECT ST_SRID(${geom_column}) FROM ${table} LIMIT 1)
        ),
        ${geom_column}
      ) as distance
  
    FROM
    ${table}
  
    ${filter ? `WHERE ${filter}` : ""}
  
    ORDER BY
      ${geom_column} <-> ST_Transform(
        st_setsrid( st_makepoint(${x}, ${y}), ${srid} ),
        (SELECT ST_SRID(${geom_column}) FROM ${table} LIMIT 1)
      )
  
    LIMIT ${limit}
    `;
}

// src/queries/query_table.ts
function query_table(table, columns, filter, group, sort, limit) {
  return `SELECT
    ${columns}
  FROM
  ${table}
  ${filter ? `WHERE ${filter}` : ""}
  ${group ? `GROUP BY ${group}` : ""}
  ${sort ? `ORDER BY ${sort}` : ""}
  ${limit ? `LIMIT ${limit}` : ""}
`;
}

// src/queries/transform_point.ts
function transform_point(point, srid) {
  const { x, y, srid: srid1 } = parsePoint(point);
  return `
  SELECT
    ST_X(
      ST_Transform(
        ST_SetSRID(
          ST_MakePoint(${x}, ${y}),
          ${srid1}
        ),
        ${srid}
      )
    ) as x,
    ST_Y(
      ST_Transform(
        ST_SetSRID(
          ST_MakePoint(${x}, ${y}),
          ${srid1}
        ),
        ${srid}
      )
    ) as y
  `;
}

// src/utils/logger.ts
var DEBUG = process.env["POSTGIS_DEBUG"] === "true";
function formatMessage(level, args) {
  const ts = (/* @__PURE__ */ new Date()).toISOString();
  return `[postgis:${level}] ${ts} ${args.map(String).join(" ")}`;
}
var logger = {
  debug(...args) {
    if (DEBUG) {
      console.debug(formatMessage("debug", args));
    }
  },
  warn(...args) {
    console.warn(formatMessage("warn", args));
  },
  error(...args) {
    console.error(formatMessage("error", args));
  }
};

// src/index.ts
var Postgis = class {
  /**
   * @param client - A connected `pg.Client` or any object with a `query(sql)` method.
   * @throws {TypeError} if `client` is missing or does not expose a `query` function.
   */
  constructor(client) {
    if (!client || typeof client.query !== "function") {
      throw new TypeError("A valid pg.Client instance is required.");
    }
    this.client = client;
  }
  // ─── Schema Introspection ────────────────────────────────────────────────
  /**
   * List all user-accessible tables in the database, joined with PostGIS
   * geometry metadata where available.
   *
   * @param options.filter - Optional SQL WHERE clause fragment.
   * @returns Array of table metadata rows.
   *
   * @example
   * ```typescript
   * const tables = await postgis.list_tables({ filter: "table_type = 'BASE TABLE'" });
   * ```
   */
  async list_tables({ filter } = {}) {
    const sql = list_tables(filter);
    return this._executeQuery(sql);
  }
  /**
   * List all columns of a given table using PostgreSQL system catalogs.
   *
   * @param table - Table name.
   * @returns Array of `{ field_name, field_type }` rows.
   *
   * @example
   * ```typescript
   * const columns = await postgis.list_columns('my_layer');
   * ```
   */
  async list_columns(table) {
    const sql = list_columns(table);
    return this._executeQuery(sql);
  }
  // ─── General Querying ────────────────────────────────────────────────────
  /**
   * Query a table with optional column selection, filtering, grouping,
   * sorting and limiting.
   *
   * @param table - Table name.
   * @param options - Query options.
   * @returns Array of result rows.
   *
   * @example
   * ```typescript
   * const rows = await postgis.query_table('parcels', {
   *   columns: 'id, name',
   *   filter: "status = 'active'",
   *   sort: 'name ASC',
   *   limit: 50,
   * });
   * ```
   */
  async query_table(table, { columns = "*", filter, group, sort, limit = 100 } = {}) {
    const sql = query_table(table, columns, filter, group, sort, limit);
    return this._executeQuery(sql);
  }
  // ─── Spatial Operations ──────────────────────────────────────────────────
  /**
   * Calculate the spatial bounding box (extent) of all geometries in a table.
   *
   * @param table - Table name.
   * @param options.geom_column - Geometry column (default `'geom'`).
   * @param options.srid - Target SRID (default `4326`).
   * @param options.filter - Optional SQL WHERE clause fragment.
   * @returns Array containing a single `{ bbox }` row.
   *
   * @example
   * ```typescript
   * const [result] = await postgis.bbox('parcels');
   * console.log(result.bbox); // BOX(lng1 lat1, lng2 lat2)
   * ```
   */
  async bbox(table, { geom_column = "geom", srid = 4326, filter } = {}) {
    const sql = bbox(table, geom_column, srid, filter);
    return this._executeQuery(sql);
  }
  /**
   * Calculate the centroid (x, y) of each geometry in a table.
   *
   * @param table - Table name.
   * @param options.force_on_surface - Use `ST_PointOnSurface` instead of `ST_Centroid` (default `false`).
   * @param options.geom_column - Geometry column (default `'geom'`).
   * @param options.srid - Target SRID (default `'4326'`).
   * @param options.filter - Optional SQL WHERE clause fragment.
   * @returns Array of `{ x, y }` rows.
   */
  async centroid(table, {
    force_on_surface = false,
    geom_column = "geom",
    srid = "4326",
    filter
  } = {}) {
    const sql = centroid(table, force_on_surface, geom_column, srid, filter);
    return this._executeQuery(sql);
  }
  /**
   * Find features from `table_from` that are within `distance` of any feature
   * in `table_to` (using `ST_DWithin`).
   *
   * @param table_from - Source table.
   * @param table_to - Target table.
   * @param options - Intersection options.
   * @returns Array of matching feature rows.
   */
  async intersect_feature(table_from, table_to, {
    columns = "*",
    distance = "0",
    geom_column_from = "geom",
    geom_column_to = "geom",
    filter,
    sort,
    limit
  } = {}) {
    const sql = intersect_feature(
      table_from,
      table_to,
      columns,
      distance,
      geom_column_from,
      geom_column_to,
      filter,
      sort,
      limit
    );
    return this._executeQuery(sql);
  }
  /**
   * Find features in `table` that are within `distance` of the given point.
   *
   * @param table - Table name.
   * @param point - Point string in `"x,y,srid"` format (e.g. `"73.5,14.9,4326"`).
   * @param options - Intersection options.
   * @returns Array of matching feature rows.
   * @throws {Error} if `point` format is invalid.
   */
  async intersect_point(table, point, {
    columns = "*",
    distance = "0",
    geom_column = "geom",
    filter,
    sort,
    limit = 10
  } = {}) {
    const sql = intersect_point(table, point, columns, distance, geom_column, filter, sort, limit);
    return this._executeQuery(sql);
  }
  // ─── Geometry Export ─────────────────────────────────────────────────────
  /**
   * Export features from `table` as a GeoJSON `FeatureCollection`.
   *
   * @param table - Table name.
   * @param options - GeoJSON export options.
   * @returns A GeoJSON `FeatureCollection`.
   *
   * @example
   * ```typescript
   * const fc = await postgis.geojson('parcels', { precision: 6 });
   * // { type: 'FeatureCollection', features: [...] }
   * ```
   */
  async geojson(table, { bounds, id_column, precision = 9, geom_column = "geom", columns, filter } = {}) {
    const sql = geojson(table, bounds, id_column, precision, geom_column, columns, filter);
    const rows = await this._executeQuery(sql);
    return {
      type: "FeatureCollection",
      features: rows.map((el) => el.geojson)
    };
  }
  /**
   * Export features from `table` as a Geobuf binary buffer.
   *
   * @param table - Table name.
   * @param options - Geobuf export options.
   * @returns A `Buffer` containing the Geobuf-encoded data.
   *
   * @example
   * ```typescript
   * const buf = await postgis.geobuf('parcels');
   * res.setHeader('Content-Type', 'application/x-protobuf');
   * res.send(buf);
   * ```
   */
  async geobuf(table, { bounds, geom_column = "geom", columns, filter } = {}) {
    const sql = geobuf(table, bounds, geom_column, columns, filter);
    const rows = await this._executeQuery(sql);
    return rows[0].st_asgeobuf;
  }
  /**
   * Generate a Mapbox Vector Tile (MVT) for the given `z/x/y` tile coordinate.
   *
   * @param table - Table name.
   * @param x - Tile X coordinate.
   * @param y - Tile Y coordinate.
   * @param z - Zoom level.
   * @param options - MVT options.
   * @returns Array containing the raw MVT binary in a `{ mvt }` row.
   *
   * @example
   * ```typescript
   * const [{ mvt: tile }] = await postgis.mvt('parcels', 0, 0, 0);
   * res.setHeader('Content-Type', 'application/vnd.mapbox-vector-tile');
   * res.send(tile);
   * ```
   */
  async mvt(table, x, y, z, { columns, id_column, geom_column = "geom", filter } = {}) {
    const sql = mvt(table, x, y, z, columns, id_column, geom_column, filter);
    return this._executeQuery(sql);
  }
  /**
   * Find the nearest features in `table` to a given point, ordered by distance.
   *
   * @param table - Table name.
   * @param point - Point string in `"x,y,srid"` format (e.g. `"73.5,14.9,4326"`).
   * @param options - Nearest options.
   * @returns Array of nearest feature rows (each includes a `distance` column).
   * @throws {Error} if `point` format is invalid.
   */
  async nearest(table, point, { columns = "*", geom_column = "geom", filter, limit = 10 } = {}) {
    const sql = nearest(table, point, columns, geom_column, filter, limit);
    return this._executeQuery(sql);
  }
  /**
   * Transform a point from one coordinate reference system to another.
   *
   * @param point - Point string in `"x,y,srid"` format (e.g. `"73.5,14.9,4326"`).
   * @param options.srid - Target SRID (default `4326`).
   * @returns Array of `{ x, y }` rows with the transformed coordinates.
   * @throws {Error} if `point` format is invalid.
   *
   * @example
   * ```typescript
   * const [pt] = await postgis.transform_point('73.5,14.9,4326', { srid: 3857 });
   * ```
   */
  async transform_point(point, { srid = 4326 } = {}) {
    const sql = transform_point(point, srid);
    return this._executeQuery(sql);
  }
  // ─── Internal ────────────────────────────────────────────────────────────
  /**
   * Execute a raw SQL string against the client and return result rows.
   * @internal
   */
  async _executeQuery(sql) {
    logger.debug("Executing query:", sql.replace(/\s+/g, " ").trim().slice(0, 120));
    try {
      const res = await this.client.query(sql);
      return res.rows ?? [];
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      logger.error("Query execution failed:", message);
      throw new Error(`Query execution failed: ${message}`);
    }
  }
};
var index_default = Postgis;

exports.Postgis = Postgis;
exports.default = index_default;
